lib/Partitioner/Partitioner.cpp

```diff
@@ -0,0 +1,324 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "glow/Partitioner/Partitioner.h"
+#include "glow/Graph/Context.h"
+#include "glow/Graph/Utils.h"
+
+#include "llvm/Support/Casting.h"
+#include "llvm/Support/raw_ostream.h"
+
+using namespace glow;
+using llvm::isa;
+
+/// A graph with BFS oder.
+struct BFSLevel {
+  /// A list of <level, nodelist> with BFS order.
+  std::vector<std::pair<int, std::vector<Node *>>> levels;
+  /// A set of visited nodes.
+  std::unordered_set<const Node *> visited;
+};
+
+Partitioner::Partitioner(Module *parent, std::vector<DeviceInfo> &devices)
+    : module_(parent), deviceInfo_(devices) {
+  // Check if all the functions are in the same structure.
+  assert(isSameFunctionFamily() && "not in the same function family");
```

Comment:
I think that we decided not to use the concept of family functions.

Reply:
@nadavrot We are not using function families for partitioning (this definitely makes sense to me) or at all (I'm not clear yet on how this can done)?

Reply:
The semantics of "family" is not well defined. You can say that some functions are related if they share constants, or something, but these are all things that you can express in the Graph  (with functions that refer to constants).